Source/WebCore/page/scrolling/ScrollAnchoringController.cpp

```diff
@@ -50,35 +53,138 @@ ScrollOffset ScrollAnchoringController::computeOffset(RenderBox& candidate)
 
-    Element* candidateAnchor = nullptr;
+            if (auto renderer = element->renderBox()) {
+                if (renderer->style().overflowAnchor() == OverflowAnchor::None)
+                    return false;
+            }
+            iterElement = iterElement->parentElement();
+        }
+
+        if (iterElement) {
+            m_anchorElement = candidateElement;
+            m_lastOffsetForAnchorElement = computeOffset(*m_anchorElement->renderBox());
+            return true;
+        }
+    }
+    return false;
+}
+
+bool ScrollAnchoringController::findPriorityCandidate(Document& document)
+{
+    // TODO: need to check if focused element is text editable
+    // TODO: need to figure out how to get element that is the current find-in-page element (look into FindController)
+    return examinePriorityCandidate(document.focusedElement());
+}
+
+CandidateExaminationResult ScrollAnchoringController::examineCandidate(Element& element)
+{
     auto layoutViewport = m_frameView.layoutViewportRect();
+    auto* document = m_frameView.frame().document();
 
-    // TODO: iterate from the current scroller not document root
-    for (auto& element : descendantsOfType<Element>(document->rootNode())) {
-        if (auto renderer = element.renderBox()) {
-            if (renderer->style().overflowAnchor() == OverflowAnchor::None || &element == document->bodyOrFrameset() || is<HTMLHtmlElement>(&element))
-                continue;
-            auto boxRect = renderer->absoluteBoundingBoxRect();
-            if (layoutViewport.intersects(boxRect))
-                candidateAnchor = &element;
-            if (layoutViewport.contains(boxRect))
-                break;
+    if (auto renderer = element.renderBox()) {
+        // TODO: we need to think about position: absolute
+        // TODO: figure out how to get scrollable area for renderer to check if it is maintaining scroll anchor
+        if (renderer->style().overflowAnchor() == OverflowAnchor::None || renderer->isStickilyPositioned() || renderer->isFixedPositioned() || renderer->isPseudoElement() || renderer->isAnonymousBlock() || (renderer->isInline() && !renderer->isAtomicInlineLevelBox()))
+            return CandidateExaminationResult::Exclude;
+        if (&element == document->bodyOrFrameset() || is<HTMLHtmlElement>(&element))
+            return CandidateExaminationResult::Skip;
+        auto boxRect = renderer->absoluteBoundingBoxRect();
+        if (!boxRect.width() || !boxRect.height())
+            return CandidateExaminationResult::Skip;
+        if (layoutViewport.contains(boxRect))
```

Comment:
Need tests with various types of zooming to ensure that `absoluteBoundingBoxRect` and `layoutViewport` share the same coordinate space in those scenarios.

Reply:
This will also need some changing in https://github.com/WebKit/WebKit/pull/17934 as we want to use the visible area for the owning scroller.